sql/range.go

```diff
@@ -429,6 +429,25 @@ func RemoveOverlappingRanges(ranges ...Range) (RangeCollection, error) {
 		return nil, nil
 	}
 
+	var firstNonNull int
+	var foundNull bool
+	for i, rang := range ranges {
+		foundNull = false
+		for _, e := range rang {
+			if e.Type() == RangeType_Null {
+				foundNull = true
+				break
+			}
+		}
+		if !foundNull {
+			firstNonNull = i
+			break
+		}
+	}
+
+	// RB-tree can't compare ranges with null base type.
+	ranges[0], ranges[firstNonNull] = ranges[firstNonNull], ranges[0]
```

Comment:
I'm not sure what's going on here. Why do we swap the first range in the slice with the position of the first non-null range? It doesn't seem like it accomplishes anything since we iterate over every range anyway, and the null and non-null distribution may potentially be random.

In addition, the note mentions that the tree can't compare ranges with the null base type, so are those not considering in the overlap removal? If so, we introduce the potential to have duplicate rows coming from integrators, which would be a major problem.

Reply:
It's pretty subtle but this change prevented duplicate rows and dropped from in testing, depending on the ranges. The RB-tree inits the type for the whole tree from the first range, i think one example iirc is: `x IS NULL OR x = 1 OR x = 2`. So the first type dictates if we can do range comparisons/intersections or not.  If the tree type is null, we can't do any numeric comparisons (because null type), and so in the above case x = 2 just gets dropped I think? They were weird bugs so I might not be remembering the exact behavior correctly.

In general there are three cases:
1) there are only nulls, and it doesn't matter what the type is.
2) there are a mix of nulls and non-nulls, in which case we need the type from one of the not nulls or else we can't overlap ranges
3) there are no nulls, so everything will work as expected

Reply:
Oh I see what you mean now. I forgot about the tree grabbing its type from the first range. With that, I think all of this logic doesn't quite belong here and should be moved to the tree initialization function, in case a tree is ever created in some other context (could be used by an integrator). Perhaps the types are passed as a slice in a separate parameter, and this loop grabs the first non-null type for each column expression, rather than the first range that contains all non-null column expressions (as all ranges may have at least one null column expression, should be a test for that if there isn't one).

Code could look something like:
```go
// NewRangeColumnExprTree creates a new RangeColumnExprTree constructed from an initial range. As the initial Range may
// contain column expressions that have a NULL type, the expected non-NULL type for each column expression is given
// separately. If all column expressions for a specific column will be NULL, then it is valid to use the NULL type.
// Returns an error if the number of column expressions do not equal the number of types, or if the Range has a length
// of zero.
func NewRangeColumnExprTree(initialRange Range, columnExprTypes []Type) (*RangeColumnExprTree, error) {
	if len(initialRange) != len(columnExprTypes) {
		return nil, fmt.Errorf("number of types given do not correspond to the number of column expressions")
	}
	if len(initialRange) == 0 {
		return nil, fmt.Errorf("a RangeColumnExprTree cannot be created from a Range of length 0")
	}

	var tree *RangeColumnExprTree
	var parent *RangeColumnExprTree
	for i, colExpr := range initialRange {
		innerTree := &RangeColumnExprTree{
			typ:  columnExprTypes[i],
			size: 1,
			root: nil,
		}
		innerTree.root = &rangeColumnExprTreeNode{
			color:         black,
			LowerBound:    colExpr.LowerBound,
			UpperBound:    colExpr.UpperBound,
			MaxUpperbound: colExpr.UpperBound,
			Inner:         nil,
			Left:          nil,
			Right:         nil,
			Parent:        nil,
		}
		if tree == nil {
			tree = innerTree
			parent = innerTree
		} else {
			parent.root.Inner = innerTree
			parent = innerTree
		}
	}
	return tree, nil
}
```

Reply:
yep this is clearer and easier 👍